src/features/island/collectibles/components/GenieLamp.tsx

```diff
@@ -1,19 +1,132 @@
+import { Revealed } from "features/game/components/Revealed";
+import { useActor } from "@xstate/react";
+import { Modal } from "react-bootstrap";
+import classNames from "classnames";
+
+import genieImg from "assets/npcs/genie.png";
+import { setImageWidth } from "lib/images";
+import { Button } from "components/ui/Button";
+import { CloseButtonPanel } from "features/game/components/CloseablePanel";
+
+interface Props {
+  id: string;
+}
+
+export const GenieLamp: React.FC<Props> = ({ id }) => {
+  const { gameService } = useContext(Context);
+  const [gameState] = useActor(gameService);
+
+  const lamps = gameState.context.state.collectibles["Genie Lamp"];
+  const lamp = lamps?.find((lamp) => lamp.id === id);
+  const rubbedCount = lamp?.rubbedCount ?? 0;
+  const wishesRemaining = 3 - rubbedCount;
+  const hasBeenRubbed = rubbedCount > 0;
+
+  const [isConfirming, setIsConfirming] = useState(false);
+  const [isRevealing, setIsRevealing] = useState(false);
+
+  // This useEffect is for the case where it is the last genie lamp wish.
+  // Genie Lamp will be deleted from the game state, so we need to make sure
+  // we do no get stuck in the revealing state.
+  useEffect(() => {
+    return () => {
+      gameService.send("CONTINUE");
+    };
+  }, []);
+
+  const rub = () => {
+    setIsConfirming(false);
+    setIsRevealing(true);
+
+    gameService.send("REVEAL", {
+      event: {
+        type: "genieLamp.rubbed",
+        id,
+        createdAt: new Date(),
+      },
+    });
+  };
 
-export const GenieLamp: React.FC = () => {
   return (
     <>
       <img
+        onClick={() => setIsConfirming(true)}
         src={genieLamp}
         style={{
           width: `${PIXEL_SCALE * 22}px`,
+          bottom: `${PIXEL_SCALE * 4}px`,
         }}
-        className="absolute"
+        className={classNames("absolute cursor-pointer hover:img-highlight", {
+          "saturate-50": hasBeenRubbed,
+        })}
         alt="Genie Lamp"
       />
+      <Modal show={isConfirming} centered onHide={() => setIsConfirming(false)}>
+        <img
+          src={genieImg}
+          className="absolute z-0"
+          style={{
+            width: `${PIXEL_SCALE * 100}px`,
+            top: `${PIXEL_SCALE * -55}px`,
+            left: `${PIXEL_SCALE * -10}px`,
+          }}
+        />
+        <CloseButtonPanel
+          className="z-10"
+          onClose={() => setIsConfirming(false)}
+          title="Ready to make a wish?"
+        >
+          <div className="flex flex-col items-center p-2">
+            <img
+              src={genieLamp}
+              alt="genieLamp"
+              className="mb-2"
+              onLoad={(e) => setImageWidth(e.currentTarget)}
+            />
+            <span
+              className="text-center text-xs mb-1"
+              style={{
+                height: "24px",
+              }}
+            >
+              {wishesRemaining} wish{wishesRemaining > 1 && "es"} remaining!
+            </span>
```

Comment:
It would also be great to mention on the first rub - "You cannot withdraw the lamp once it has been rubbed".

Also a "read more" links to the docs would be amazing 🙏 Our players will have lots of questions around this